### PR TITLE
Fix ConcretizationTypeError in dot_product_attention with traced masks

### PR DESCRIPTION
**Pull request by Abhinavexists**
Fix: #21808 

using `dot_product_attention` with a mask inside JAX control flow operations (e.g., `while_loop`, `scan`) on TPU, the code raised a `ConcretizationTypeError`

Added a traced mask detection check before attempting to use TPU flash attention

### CODE REVIEW

File: keras/src/backend/jax/nn.py [hunk 1]
```diff
@@ -1276,6 +1276,10 @@ def dot_product_attention(
         # use flash attention
         _can_use_flash_attention(query, key, value, bias, raise_error=True)
 
+    # On TPU, traced masks cause ConcretizationTypeError with flash attention.
+    if is_tpu and mask is not None and isinstance(mask, jax.core.Tracer):
+        flash_attention = False
```
Comment (hertschuh) [inline keras/src/backend/jax/nn.py]: Because on JAX we compile by default, this means that `flash_attention` is now always disabled. The whole flash attention implementation was for TPU specifically. So I don't think this is the right fix.

Am I misunderstanding something?

Comment (Abhinavexists) [reply keras/src/backend/jax/nn.py]: Hi @hertschuh 

You're absolutely right to question this - I need to clarify.

The key distinction is that `mask` being a `Tracer` doesn't happen during normal compilation. In typical usage, the mask is a concrete array that gets traced, but the mask *value itself* isn't a Tracer.

However, when `dot_product_attention` is called inside JAX control flow primitives like `scan` or `while_loop`, the mask becomes a Tracer because these primitives abstract over the loop-carried values. This is when flash attention fails with ConcretizationTypeError.

So the check `isinstance(mask, jax.core.Tracer)` should only trigger in the specific case where we're inside control flow operations, not during normal compilation.

That said, I want to verify this is actually the case. ?

Comment (Abhinavexists) [reply keras/src/backend/jax/nn.py]: @hertschuh You're absolutely right! Thank you for catching that.

My initial fix using `isinstance(mask, jax.core.Tracer)` was too broad and would have disabled flash attention even during normal JIT compilation, not just inside control flow operations.

I've revised the fix to instead:

1. **Removed the Tracer check**
2. **Improved the exception handling** to specifically catch `jax.errors.ConcretizationTypeError`

The fix now correctly handles traced masks without disabling flash attention for normal use cases.

File: keras/src/backend/jax/nn.py [hunk 2]
```diff
@@ -1353,7 +1353,10 @@ def dot_product_attention(
             )
             # Transpose output back to Keras layout
             return jnp.transpose(output, axes=(0, 2, 1, 3))
-        except Exception:
+        except jax.errors.ConcretizationTypeError:
+            # Mask is traced
+            # Fall back to native attention
+            flash_attention = False
```
Comment (hertschuh) [inline keras/src/backend/jax/nn.py]: `flash_attention = False`

This is already line 1364, remove.

**Review by hertschuh — APPROVED** [review]

**Review by hertschuh — COMMENTED** [review]
So the TPU tests fail.

https://github.com/keras-team/keras/actions/runs/19870652602/job/56945198925?pr=21853

It look like the original code was correct. It's not just `ConcretizationTypeError`s that can happen. Other types of errors can happen and we need to turn off splash attention.

**Review by divyashreepathihalli — APPROVED** [review]
Thank you for the fix!

**Comment by hertschuh** [comment]
Closing in favor of https://github.com/keras-team/keras/pull/21891 which adds unit tests.
